### PR TITLE
[NT-1101] Vertical Payment Methods

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

Replaces our horizontal payment methods with a vertical list.

# 🤔 Why

The horizontal payment methods didn't provide an ideal user experience and, as part of other improvements to the pledge view, we are changing this to a vertical list.

# 🛠 How

- Some refactoring of `PledgePaymentMethodsViewController`.
- Replaced the `UIScrollView` that we were using before with a `UITableView` which sizes itself to its contents.
- Refactored the `PledgeCreditCardView` and renamed it to `PledgePaymentMethodCell` along with related files.
  - **Note:** The renaming of this file unfortunately shows the entire file as new. This is probably fine considering that most of the view was refactored.

# 👀 See

| Before 🐛 | After 🦋 |
| --- | --- |
| image | image |

# ✅ Acceptance criteria

Navigate to the pledge view for a project.

- [ ] The payment methods should be presented as a vertical list.
- [ ] Unavailable payment methods should be ordered last.
- [ ] Unavailable payment methods should be slightly faded and display a label with the reason for them not being available.
- [ ] You should not be able to select unavailable payment methods.
- [ ] No regressions to the behaviour of payment methods in any other contexts of the pledge view.

# ⏰ TODO

- [ ] The add card button cell should not resize with dynamic type because it's a button.

### CODE REVIEW

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  func configureWith(value _: Void) {}
+}
+
+// MARK: - Styles
+
+private let addButtonStyle: ButtonStyle = { button in
+  button
+    |> UIButton.lens.title(for: .normal) %~ { _ in
+      localizedString(
+        key: "New_payment_method",
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift:64]: Note to self to add this string for translation or check if we have it.

File: Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift [hunk 2]
```diff
@@ -85,43 +60,31 @@ final class PledgePaymentMethodsViewController: UIViewController {
     _ = self.view
       |> checkoutBackgroundStyle
 
-    _ = self.cardsStackView
-      |> cardsStackViewStyle
-
-    _ = self.applePayButton
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift:91]: Removal of views in this PR to reduce conflicts with the work in #1151 and #1154.

File: KsApi/models/templates/GraphCreditCardTemplate.swift [hunk 3]
```diff
@@ -65,8 +65,8 @@ extension GraphUserCreditCard {
       GraphUserCreditCard.diners,
       GraphUserCreditCard.jcb,
       GraphUserCreditCard.discover,
-      GraphUserCreditCard.generic,
-      GraphUserCreditCard.unionPay
+      GraphUserCreditCard.unionPay,
+      GraphUserCreditCard.generic
```
Comment (justinswart) [inline KsApi/models/templates/GraphCreditCardTemplate.swift:69]: Unavailable cards are ordered last now.

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Kickstarter-iOS/DataSources/PledgePaymentMethodsDataSourceTests.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        isEnabled: false,
+        isSelected: false,
+        projectCountry: "Country 2"
+      )
+    ]
+
+    self.dataSource.load(cellData)
+
+    XCTAssertEqual(2, self.dataSource.numberOfSections(in: self.tableView))
+    XCTAssertEqual(2, self.dataSource.numberOfItems(in: 0))
```
Comment (ifbarrera) [inline Kickstarter-iOS/DataSources/PledgePaymentMethodsDataSourceTests.swift:29]: Maybe also an assertion for the `self.dataSource.numberOfItems(in: 1)`?

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+private let labelsStackViewStyle: StackViewStyle = { stackView in
+  stackView
+    |> \.axis .~ .vertical
+    |> \.spacing .~ Styles.grid(1)
+}
+
+private let lastFourLabelStyle: LabelStyle = { label in
+  label
+    |> checkoutTitleLabelStyle
+    |> \.font .~ UIFont.ksr_caption1(size: 14).bolded
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift:207]: Is it possible to use another font style that doesn't require a size override? Maybe `ksr_subhead`?

File: Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift [hunk 6]
```diff
@@ -85,43 +60,31 @@ final class PledgePaymentMethodsViewController: UIViewController {
...
-      |> verticalStackViewStyle
-      |> checkoutSubStackViewStyle
-
-    _ = self.applePaySectionStackView
-      |> applePaySectionStackViewStyle
-
-    _ = self.storedPaymentMethodsTitleLabel
-      |> storedPaymentMethodsTitleLabelStyle
+    _ = self.tableView
+      |> \.separatorInset .~ .zero
```
Comment (ifbarrera) [inline Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift:64]: Can this not go in the `init` closure on line 19?

File: Library/UIView+AutoLayout.swift [hunk 7]
```diff
@@ -45,6 +45,24 @@ public func ksr_addLayoutGuideToView() -> ((UILayoutGuide, UIView) -> (UILayoutG
   }
 }
 
+public func ksr_centerInParent(priority: UILayoutPriority = .required)
```
Comment (ifbarrera) [inline Library/UIView+AutoLayout.swift:48]: How is this different from `ksr_constrainViewToCenterInParent` on line 26? 🤔

File: Library/ViewModels/PledgePaymentMethodCellViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    self.checkmarkImageHidden = cardTypeIsAvailable.negate()
+
+    self.lastFourLabelTextColor = cardTypeIsAvailable.map {
+      $0 ? .ksr_soft_black : .ksr_dark_grey_500
+    }
+  }
+
+  fileprivate let pledgeCreditCardValueProperty = MutableProperty<PledgePaymentMethodCellData?>(nil)
+  public func configureWith(value: PledgePaymentMethodCellData) {
+    self.pledgeCreditCardValueProperty.value = value
```
Comment (ifbarrera) [inline Library/ViewModels/PledgePaymentMethodCellViewModel.swift:126]: `configureWithValueProperty`?

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift [hunk 9]
```diff
@@ -0,0 +1,76 @@
+import KsApi
+import Library
+import Prelude
+import UIKit
+
+final class PledgePaymentMethodAddCell: UITableViewCell, ValueCell {
+  // MARK: - Properties
+
+  private lazy var selectionView: UIView = { UIView(frame: .zero) |> \.backgroundColor .~ .ksr_grey_200 }()
```
Comment (Scollaco) [inline Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift:9]: I personally prefer to add the style here too, but in this case, shouldn't we a `selectionViewStyle` for consistency?

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+  private lazy var addButton: UIButton = {
+    UIButton(type: .custom)
+      |> UIButton.lens.titleLabel.font .~ UIFont.boldSystemFont(ofSize: 15)
```
Comment (Scollaco) [inline Kickstarter-iOS/Views/Cells/PledgePaymentMethodAddCell.swift:12]: Same here

File: Kickstarter-iOS/DataSources/PledgePaymentMethodsDataSourceTests.swift (hunk 4 above)
Comment (justinswart) [reply Kickstarter-iOS/DataSources/PledgePaymentMethodsDataSourceTests.swift:29]: This is actually addressed in the follow-up PR: https://github.com/kickstarter/ios-oss/pull/1156/files#diff-24128a4cb3c0c847303e6056c2ace3ea

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift (hunk 5 above)
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift:207]: I can try that.

File: Library/UIView+AutoLayout.swift (hunk 7 above)
Comment (justinswart) [reply Library/UIView+AutoLayout.swift:48]: 😬 just clowning about.

File: Library/ViewModels/PledgePaymentMethodCellViewModel.swift (hunk 8 above)
Comment (justinswart) [reply Library/ViewModels/PledgePaymentMethodCellViewModel.swift:126]: 😑

File: Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift [hunk 11]
```diff
@@ -85,43 +60,31 @@ final class PledgePaymentMethodsViewController: UIViewController {
...
 
   override func bindViewModel() {
     super.bindViewModel()
 
-    self.viewModel.outputs.reloadPaymentMethodsAndSelectCard
+    self.viewModel.outputs.reloadPaymentMethods
       .observeForUI()
-      .observeValues { [weak self] cardValues, selectedCard in
+      .observeValues { [weak self] cards, selectedCard, shouldReload in
         guard let self = self else { return }
```
[the remark this answers is not in the document]
Comment (justinswart) [reply Kickstarter-iOS/Views/Controllers/PledgePaymentMethodsViewController.swift:76]: I thin it just means we don't have to use `self?.` in the references below?

File: Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift (hunk 5 above)
Comment (justinswart) [reply Kickstarter-iOS/Views/Cells/PledgePaymentMethodCell.swift:207]: It increases kerning and size a little, will check in with @colleenmacd to see if we can switch it out.

**Comment by Scollaco** [comment]
Thanks for addressing the feedbacks, Justin. I noticed one more thing though - When dismissing the AddNewCardViewController, the bottom view can cover part of the tableView:

vertical-payment-methods

**Comment by justinswart** [comment]
> When dismissing the AddNewCardViewController, the bottom view can cover part of the tableView

@Scollaco interesting! I'll take a look 👍

**Pull request merged**